src/Storages/RocksDB/EmbeddedRocksDBBlockInputStream.cpp

```diff
@@ -0,0 +1,68 @@
+#include <DataStreams/IBlockInputStream.h>
+#include <Interpreters/Context.h>
+#include <Storages/RocksDB/StorageEmbeddedRocksDB.h>
+#include <Storages/RocksDB/EmbeddedRocksDBBlockInputStream.h>
+
+#include <rocksdb/db.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ROCKSDB_ERROR;
+}
+
+EmbeddedRocksDBBlockInputStream::EmbeddedRocksDBBlockInputStream(
+            StorageEmbeddedRocksDB & storage_,
+            const StorageMetadataPtr & metadata_snapshot_,
+            size_t max_block_size_)
+            : storage(storage_)
+            , metadata_snapshot(metadata_snapshot_)
+            , max_block_size(max_block_size_)
+{
+    sample_block = metadata_snapshot->getSampleBlock();
+    primary_key_pos = sample_block.getPositionByName(storage.primary_key);
+}
+
+Block EmbeddedRocksDBBlockInputStream::readImpl()
+{
+    if (finished)
+        return {};
+
+    if (!iterator)
+    {
+        iterator = std::unique_ptr<rocksdb::Iterator>(storage.rocksdb_ptr->NewIterator(rocksdb::ReadOptions()));
+        iterator->SeekToFirst();
+    }
+
+    MutableColumns columns = sample_block.cloneEmptyColumns();
+
+    size_t rows = 0;
+    for (; iterator->Valid(); iterator->Next())
```

Comment:
```suggestion
    for (size_t rows = 0; iterator->Valid() && rows < max_block_size; iterator->Next(); ++rows)
```

Will make stop condition clear and fix the bug with the iteration advance:

```sql
SELECT WatchID
FROM default.hits_rocksdb
LIMIT 3
SETTINGS max_block_size = 1

Query id: ee083ee9-a725-4bba-8686-db213792e387

┌─────────────WatchID─┐
│ 5715634532485893888 │
└─────────────────────┘
┌─────────────WatchID─┐
│ 5715634532485893888 │
└─────────────────────┘
┌─────────────WatchID─┐
│ 5715634532485893888 │
└─────────────────────┘
```

Reply:
good catch.